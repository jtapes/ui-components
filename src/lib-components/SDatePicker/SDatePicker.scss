@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

$default-color: $color--font;
$primary-color: $color--primary;

@import '~vue2-datepicker/scss/var.scss';
@import '~vue2-datepicker/scss/icon.scss';
@import '~vue2-datepicker/scss/btn.scss';
@import '~vue2-datepicker/scss/scrollbar.scss';
@import '~vue2-datepicker/scss/animation.scss';

$namespace: 'mx';

/* prettier-ignore */
.#{$namespace}-input-wrapper .s-input--readonly.s-input--rounded .s-input__inner {
  color: $color--font;
  background-color: $color--white;
}

.s-date-picker {
  width: 100%;
  cursor: pointer;

  &.s-date-picker--focused {
    &:not(.s-date-picker--range) {
      .s-input {
        border-color: transparent;
        border-radius: $size--24 $size--24 0 0;
        box-shadow: 0 0 0 3px $color--outline;
      }
    }

    &.s-date-picker--range {
      .#{$namespace}-input-wrapper {
        border-radius: $size--24 $size--24 0 0;
        box-shadow: 0 0 0 3px $color--outline;

        &:hover {
          border-color: transparent;
        }
      }
    }
  }

  &.s-date-picker--range {
    .#{$namespace}-input-wrapper {
      display: flex;
      border-radius: $size--24;
      padding: 0 $size--16;
      box-shadow: inset 0 0 0 1px $color--border;
      background-color: $color--white;

      &:hover {
        border-color: $color--primary-accent;
      }
    }

    .s-input {
      border-color: transparent;
      padding-top: 0;
      width: 112px;
      background: transparent;

      &__prepend {
        white-space: nowrap;
      }

      &__inner {
        @include text-sm;

        padding: 0;
      }

      &:focus-within {
        box-shadow: none;
      }

      &:hover {
        box-shadow: none;
      }

      &.s-input--focused {
        border-radius: 0;
        box-shadow: none;
      }
    }
  }
}

.#{$namespace}-datepicker {
  position: relative;
  display: inline-block;
  width: 100%;

  svg {
    vertical-align: -0.15em;
    overflow: hidden;
    width: 1em;
    height: 1em;
    fill: currentColor;
  }
}

.#{$namespace}-datepicker-inline {
  width: auto;
}

.#{$namespace}-input-wrapper {
  position: relative;
}

.#{$namespace}-icon-calendar {
  position: absolute;
  right: $size--12;
  top: 50%;
  vertical-align: middle;
  font-size: $size--32;
  line-height: 1;
  color: rgba(0, 0, 0, 0.5);
  transform: translateY(-50%);
}

.#{$namespace}-datepicker-main {
  color: $default-color;
  background-color: #fff;
}

.#{$namespace}-datepicker-popup {
  position: absolute;
  z-index: $popup-z-index;
  margin-top: -3px;
  border-top: 1px solid $border-color;
  border-radius: 0 0 $size--24 $size--24;
  box-shadow: 0 3px 0 #99b3d8, 3px 3px 0 #99b3d8, -3px 3px 0 #99b3d8;

  .#{$namespace}-calendar-panel-date + .#{$namespace}-calendar-panel-date {
    display: none;
  }
}

.#{$namespace}-datepicker-sidebar {
  float: left;
  overflow: auto;
  padding: $size--8;
  width: $sidebar-margin-left;
  box-sizing: border-box;
}

.#{$namespace}-datepicker-sidebar + .#{$namespace}-datepicker-content {
  margin-left: $sidebar-margin-left;
  border-left: 1px solid $border-color;
}

.#{$namespace}-datepicker-body {
  position: relative;
  user-select: none;
}

.#{$namespace}-btn-shortcut {
  display: block;
  padding: 0 $size--8;
  line-height: $size--24;
}

.#{$namespace}-range-wrapper {
  display: flex;
}

.#{$namespace}-datepicker-header {
  border-bottom: 1px solid $border-color;
  padding: $size--8;
}

.#{$namespace}-datepicker-footer {
  border-top: 1px solid $border-color;
  padding: $size--8;
  text-align: right;
}

.#{$namespace}-calendar {
  padding: $size--16 $size--8;
  box-sizing: border-box;
}

.#{$namespace}-calendar-header {
  overflow: hidden;
  padding: 0 $size--8;
  height: $size--32;
  box-sizing: border-box;
  line-height: $size--32;
  text-align: center;
}

.#{$namespace}-btn-icon-left,
.#{$namespace}-btn-icon-double-left {
  float: left;
}

.#{$namespace}-btn-icon-right,
.#{$namespace}-btn-icon-double-right {
  float: right;
}

.#{$namespace}-icon-right, .#{$namespace}-icon-left {
  &::before {
    border: 0;
    width: $size--16;
    height: $size--16;
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAMAAABg3Am1AAAARVBMVEUAAAAAMHAAMGgAMHAAMGoALW0AMGwALmsAMGsALW0AMG0ALmwAMGwALm0AL2sAL20AL2wAL2wAL2wAL2sAL20AL2wAL2wfSl78AAAAFnRSTlMAECAgMD9AX19gYH+Aj5+fr7/P39/vQLAKjwAAALJJREFUeNrt04USxCAQA9CFunv//1OvLeHG2Jmca8ZfqsDKp+afrrYXebSuU005Uq5bppx2GTZHRbmYavYV5VuianUZYsJd1aFqre4vucUvTG0p35KMfmE431JgLXvSz9XAOSoUhCNa8Tk3mBoF5RsvWD7OywUHh/PUz0xGeTo4XirDeBRi3Y0/9Q1YdyyB48GK7gFOhPESSzBuzHgfXDHdpcISsC5mAXO+pWutaP6s/HMCkhsjx+D0EuYAAAAASUVORK5CYII=');
    background-size: cover;
  }
}

.#{$namespace}-icon-right {
  &::before {
    left: 0;
    transform: none;
  }
}

.#{$namespace}-icon-left {
  &::before {
    transform: rotate(180deg);
  }
}

.#{$namespace}-icon-double-right, .#{$namespace}-icon-double-left {
  &::before {
    border: 0;
    width: $size--16;
    height: $size--16;
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAMAAABg3Am1AAAAUVBMVEUAAAAAMHAAMGgAMHAALW0AMGwALWwAMGwALmsAMGsALW0AMG0ALmwAMGsALmwAMGwALm0AMGwAL2sAL20AL2wAL2wAL2wAL2sAL20AL2wAL2zVb8n+AAAAGnRSTlMAECAgP0BQUF9fYGBvcH+Aj5Cfn6+/z9/f70DZzEIAAACTSURBVHja7ZQ1AsNQDEPtMDN93f+eRXe3plLerBcwyUdyMk8RlY8B9IxS48be+4UVeCil26gOWuloJe7wZI3cymzK9GnKYn/fu5XE2rII2ZZVCCpWkO8XdKQE7QMYoQ02Hb54/lqMzBVPbSxCw8U7pUZ7UKI0wBJR8TXxXTIuLptVshAn3bM0xOgEq6SbYVA5+QKuzuAVSLY9ek8AAAAASUVORK5CYII=');
    background-size: cover;
    transform: none;
  }

  &::after {
    display: none;
  }
}

.#{$namespace}-icon-double-right {
  &::before {
    transform: none;
  }
}

.#{$namespace}-icon-double-left {
  &::before {
    transform: rotate(180deg);
  }
}

.#{$namespace}-calendar-header-label {
  @include text-sm;
}

.#{$namespace}-calendar-decade-separator {
  margin: 0 $size--4;

  &::after {
    content: '~';
  }
}

.#{$namespace}-calendar-content {
  position: relative;
  height: 224px;
  box-sizing: border-box;

  .cell {
    cursor: pointer;

    div {
      display: flex;
      align-items: center;
      justify-content: center;
      margin: 0 auto;
      width: $size--32;
      height: $size--32;
      min-width: $size--32;
      min-height: $size--32;
    }

    &:hover {
      border-radius: $size--4;
      color: $calendar-hover-color;
      background-color: $calendar-hover-background-color;
    }

    &.active {
      border-radius: $size--4;
      color: $calendar-active-color;
      background-color: $calendar-active-background-color;
    }

    &.in-range,
    &.hover-in-range {
      color: $calendar-in-range-color;
      background-color: rgba($color--complimentary-bg, 0.5);
    }

    &.disabled {
      color: $disabled-color;
      background-color: $disabled-background-color;
      cursor: not-allowed;
    }
  }
}

.#{$namespace}-calendar-week-mode {
  .#{$namespace}-date-row {
    cursor: pointer;

    .cell {
      &:hover {
        color: inherit;
        background-color: transparent;
      }

      &.active {
        color: inherit;
        background-color: transparent;
      }
    }

    &:hover {
      background-color: $calendar-hover-background-color;
    }

    &.#{$namespace}-active-week {
      background-color: $calendar-in-range-background-color;
    }
  }
}

.#{$namespace}-week-number {
  opacity: 0.5;
}

.#{$namespace}-table {
  table-layout: fixed;
  border-collapse: separate;
  border-spacing: 0;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  text-align: center;

  th {
    vertical-align: middle;
    padding: 0;
    font-weight: 500;
  }

  td {
    vertical-align: middle;
    padding: 0;
  }
}

.#{$namespace}-table-date {
  td,
  th {
    @include text-xs;

    height: $size--32;
  }

  .today {
    color: $today-color;
  }

  .cell.not-current-month {
    color: #ccc;
    background: none; // cover the in-range style
  }
}

.#{$namespace}-datepicker-range {
  input {
    padding-left: $size--32;
  }

  .#{$namespace}-calendar + .#{$namespace}-calendar {
    display: none;
  }
}
