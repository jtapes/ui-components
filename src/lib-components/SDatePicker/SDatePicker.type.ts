export type SDatePickerValueType = Date | [Date, Date] | string | [string, string]
