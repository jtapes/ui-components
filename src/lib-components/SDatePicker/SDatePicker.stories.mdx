import { action } from '@storybook/addon-actions'
import { withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SDatePicker from '@/lib-components/SDatePicker/SDatePicker.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

<Meta
  title="DESIGN SYSTEM/Atom/DatePicker"
  component={SDatePicker}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    value: {
      control: 'date',
      defaultValue: '',
      description: 'Дата',
      type: 'Date'
    },
    placeholder : {
      control: 'text',
      defaultValue: 'Выберите дату',
      description: 'Плейсхолдер',
      type: 'String'
    },
    valueStart: {
      control: 'date',
      defaultValue: '',
      description: 'Дата начала',
      type: 'Date'
    },
    valueEnd: {
      control: 'date',
      defaultValue: '',
      description: 'Дата завершения',
      type: 'Date'
    }
  }}
  parameters={{
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>



export const Template = (args, {argTypes}) => ({
  props: Object.keys(argTypes),
  components: {SDatePicker, SvgIcon},
  template: `<div style="width: 300px; height: 300px;"><s-date-picker v-model="currentValue" @change-date="changeValue" :placeholder="placeholder" /></div>`,
  data() {
    return {
      currentValue: ''
    }
  },
  watch: {
    value() {
      this.currentValue = new Date(this.value)
    }
  },
  methods: {
    changeValue(v) {
      action('change')(v)
      this.currentValue = v
    },
  }
})

export const TemplateRange = (args, {argTypes}) => ({
  props: Object.keys(argTypes),
  components: {SDatePicker, SvgIcon},
  template: `<div style="width: 300px; height: 100px;"><s-date-picker :value="currentValue" @change-date="changeValue" :range="true" placeholder="Выберите даты" :appendToBody="true" /></div>`,
  data() {
    return {
      currentValue: ['','']
    }
  },
  watch: {
    valueStart() {
      this.currentValue = [new Date(this.valueStart), new Date(this.valueEnd)]
    },
    valueEnd() {
      this.currentValue = [new Date(this.valueStart), new Date(this.valueEnd)]
    }
  },
  methods: {
    changeValue(v) {
      action('change')(v)
      this.currentValue = v
    },
  }
})

<Canvas withToolbar withSource="none">
  <Story name="SDatePicker">
    {Template.bind({})}
  </Story>
  <Story name="SDatePickerRange">
    {TemplateRange.bind({})}
  </Story>
</Canvas>

<Description of={SDatePicker} />
<ArgsTable of={SDatePicker} />
<Source
  language="html"
  code={`<s-date-picker :value="currentValue" @change-date="changeValue" :placeholder="placeholder" />`}
/>
