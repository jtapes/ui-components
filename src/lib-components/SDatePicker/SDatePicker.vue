<template>
  <div
    ref="wrapper"
    class="s-date-picker"
    :class="{
      's-date-picker--range': range,
      's-date-picker--focused': isOpen
    }"
  >
    <date-picker
      :value="value"
      v-bind="$attrs"
      :append-to-body="appendToBody"
      :open.sync="isOpen"
      :range="range"
      :clearable="false"
      :popup-style="{
        width: appendToBody ? appendedToBodyWidth : ''
      }"
      v-on="$listeners"
      @change="change"
    >
      <template #input>
        <template v-if="range">
          <s-input
            :value="dateString[0]"
            :state="state"
            :rounded="true"
            :placeholder="dateString[0] ? '' : placeholder"
            :class="{
              'w-100': !dateString[0]
            }"
          >
            <template v-if="dateString[0]" #prepend>
              <div class="mr-8 color-font-secondary text-sm">
                c
              </div>
            </template>
          </s-input>
          <s-input
            v-if="dateString[1]"
            :value="dateString[1]"
            :state="state"
            :rounded="false"
          >
            <template #prepend>
              <div class="mr-8 color-font-secondary text-sm">
                по
              </div>
            </template>
          </s-input>
        </template>
        <template v-else>
          <s-input
            :value="dateString"
            :placeholder="placeholder"
            :state="state"
          />
        </template>
      </template>
      <template #icon-calendar>
        <svg-icon :name="'calendar-24-32'" class="icon-24" />
      </template>
    </date-picker>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Model, Prop } from 'vue-property-decorator'
import DatePicker from 'vue2-datepicker'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import SInput from '@/lib-components/SInput/SInput.vue'
import { format, parse } from '@/plugins/date'
import 'vue2-datepicker/locale/ru'
import { SDatePickerValueType } from '@/lib-components/SDatePicker/SDatePicker.type'

@Component({ components: { DatePicker, SInput, SvgIcon } })
export default class SDatePicker extends Vue {
  @Model('change-date')
  value!: SDatePickerValueType

  /** Формат даты */
  @Prop({ type: String, default: 'dd.MM.yyyy' })
  format!: string

  /** Плейсхолдер */
  @Prop({ type: String, default: '' })
  placeholder!: string

  /** Состояние инпута */
  @Prop(String)
  state?: string

  /** Возможность выбора интервала дат */
  @Prop(Boolean)
  range?: boolean

  /** Свойство выноса дропдауна календаря в body документа */
  @Prop({ type: Boolean, default: false })
  appendToBody!: boolean

  isOpen = false
  appendedToBodyWidth = ''

  get date(): Date | [Date, Date] {
    if (typeof this.value === 'string' && this.value.length) {
      return parse(this.value, this.format, new Date())
    } else if (
      Array.isArray(this.value) &&
      typeof this.value[0] === 'string' &&
      typeof this.value[1] === 'string' &&
      this.value[0].length && this.value[1].length
    ) {
      return [
        parse(this.value[0], this.format, new Date()),
        parse(this.value[1], this.format, new Date()),
      ]
    } else {
      return this.value as Date | [Date, Date]
    }
  }

  get dateString(): string | [string, string] | undefined {
    if (this.date) {
      if (!Array.isArray(this.date)) {
        return format(this.date, this.format)
      } else {
        const dateStart = this.date[0] ? format(this.date[0], this.format) : ''
        const dateEnd = this.date[1] ? format(this.date[1], this.format) : ''
        return [dateStart, dateEnd]
      }
    }
  }

  change(date: Date | [Date, Date]): void {
    let dateString: string | [string, string] | undefined
    if (typeof this.value === 'string' && !Array.isArray(date)) {
      dateString = format(date, this.format)
    } else if (
      Array.isArray(this.value) &&
      typeof this.value[0] === 'string' &&
      typeof this.value[1] === 'string' &&
      Array.isArray(date)
    ) {
      dateString = [format(date[0], this.format), format(date[1], this.format)]
    }
    this.$emit('change-date', date || dateString)
  }

  prepareDropdown(): void {
    if (!this.$isServer) {
      this.$nextTick(async () => {
        if (this.$refs.wrapper) {
          const slotRect = (this.$refs.wrapper as Element).getBoundingClientRect()
          this.appendedToBodyWidth = `${slotRect.width}px`
        }
      })
    }
  }

  mounted(): void {
    window.addEventListener('resize', this.prepareDropdown, true)
    if (this.appendToBody) {
      this.prepareDropdown()
    }
  }

  beforeDestroy(): void {
    document.removeEventListener('resize', this.prepareDropdown, true)
  }
}
</script>
