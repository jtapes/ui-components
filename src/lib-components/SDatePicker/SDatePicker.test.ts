import { mount } from '@vue/test-utils'
import SDatePicker from '@/lib-components/SDatePicker/SDatePicker.vue'

describe('SDatePicker', () => {
  test('SDatePicker', () => {
    const wrapper: any = mount(SDatePicker, {
      propsData: {
        value: new Date('2021-08-23'),
      },
    })
    expect(wrapper.find('input').element.value).toBe('23.08.2021')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('SDatePicker range', () => {
    const wrapper: any = mount(SDatePicker, {
      propsData: {
        value: [new Date('2021-08-23'), new Date('2021-09-24')],
        range: true,
      },
    })
    expect(wrapper.findAll('input').wrappers[0].element.value).toBe(
      '23.08.2021'
    )
    expect(wrapper.findAll('input').wrappers[1].element.value).toBe(
      '24.09.2021'
    )
    expect(wrapper.element).toMatchSnapshot()
  })
})
