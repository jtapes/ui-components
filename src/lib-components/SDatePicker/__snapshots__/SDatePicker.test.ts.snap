// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SDatePicker SDatePicker 1`] = `
<div
  class="s-date-picker"
>
  <div
    class="mx-datepicker"
  >
    <div
      class="mx-input-wrapper"
    >
      <div
        class="s-input s-input--text s-input--rounded"
      >
        <!---->
         
        <input
          class="s-input__inner"
          placeholder=""
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
      <i
        class="mx-icon-calendar"
      >
        <svg
          class="svg-icon icon-24 svg-icon--calendar-24-32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            href="#calendar-24-32"
            xlink:href="#calendar-24-32"
          />
        </svg>
      </i>
    </div>
    <transition-stub
      name="mx-zoom-in-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`SDatePicker SDatePicker range 1`] = `
<div
  class="s-date-picker s-date-picker--range"
>
  <div
    class="mx-datepicker mx-datepicker-range"
  >
    <div
      class="mx-input-wrapper"
    >
      <div
        class="s-input s-input--text s-input--rounded"
      >
        <div
          class="s-input__prepend"
        >
          <div
            class="mr-8 color-font-secondary text-sm"
          >
            
              c
            
          </div>
        </div>
         
        <input
          class="s-input__inner s-input__inner--prepended"
          placeholder=""
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="s-input s-input--text"
      >
        <div
          class="s-input__prepend"
        >
          <div
            class="mr-8 color-font-secondary text-sm"
          >
            
              по
            
          </div>
        </div>
         
        <input
          class="s-input__inner s-input__inner--prepended"
          placeholder=""
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
      <i
        class="mx-icon-calendar"
      >
        <svg
          class="svg-icon icon-24 svg-icon--calendar-24-32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            href="#calendar-24-32"
            xlink:href="#calendar-24-32"
          />
        </svg>
      </i>
    </div>
    <transition-stub
      name="mx-zoom-in-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;
